Update memory map

mchack-work opened the pull request:
First stab at updating the memory map.

Some "XXX" marks for when I'm not sure.

If there are real problems, and you have commit acces, just commit your corrections and we can squash later. Otherwise please add your comments.

Closes #43

secworks commented on hugo/content/memory/_index.md:
```diff
@@ -35,69 +33,74 @@ The first 2 bits in a 32-bit address.
+| MMIO TOUCH  | 0xc4                            |
+| MMIO FW_RAM | 0xd0                            |
+| MMIO QEMU   | 0xfe Not used in real hardware! |
+| MMIO TK1    | 0xff                            |
```
I'm sort of not fancy of the "MMIO" part. It is not I/O. It is internal functions. I know that it has become some sort of standard to call anything strictly not memory as MMIO. But do we gain anything by adding MMIO on front of everything here? This table explains what functionality is where at a high level. RAM is a function just as well as FW_RAM.

Side note: We could implement RAM and ROM using registers, would that mean that we should att MMIO there too?

mchack-work replied:
I removed MMIO completely from this table.

secworks commented on hugo/content/memory/_index.md:
```diff
@@ -35,69 +33,74 @@ The first 2 bits in a 32-bit address.
-| MMIO TK1    | 0xff     |
+| *name*      | *prefix*                        |
+|-------------|---------------------------------|
+| ROM         | 0x00                            |
```
Since we talk about 'FW_RAM', should we call this 'FW_ROM'?

secworks commented on hugo/content/memory/_index.md:
```diff
@@ -35,69 +33,74 @@ The first 2 bits in a 32-bit address.
-[^3]: Each word of the UDS can only be read *once* after TKey has started.
+All the constants in `tk1_mem.h` are prefixed with `TK1_MMIO_` but to
+save space we present them here without the prefix.
+
```
Or drop MMIO.

mchack-work replied:
Is this a suggestion to drop the `MMIO` prefix in `tk1_mem.h` in tillitis-key1 and tkey-libs?

secworks replied:
Yes. Dropping the MMIO_ prefix here, but not in header files etc means that we create a confusion, even if we have that note.

As stated before, I don't fancy the MMIO term (even if I know it probably is a lost battle). But to me it does not make sense calling things that clearly are not I/O MMIO. The API names in the cores does not have MMIO_-prefix. It has been added later. Which I think cause confusion too. (If we are to be pedantic about ASLR, we should be it with MMIO too.)

Either have MMIO_ everywhere or nowhere, even in this document though it is considered cumbersome. And This would also mean that we should update the whole API space for all cores with MMIO_.

secworks commented on hugo/content/memory/_index.md:
```diff
@@ -35,69 +33,74 @@ The first 2 bits in a 32-bit address.
+| `QEMU_DEBUG`        | 0xfe00100      | w    | w     | Debug console (only in QEMU). XXX Which byte? LSB?                                                                                                                                                                    |
+| `TK1_NAME0`         | 0xff000000     | r    | r     | ID of core/stick, first part, default "tk1".                                                                                                                                                                          |
+| `TK1_NAME1`         | 0xff000004     | r    | r     | ID of core/stick, second part, default "mkdf".                                                                                                                                                                        |
+| `TK1_VERSION`       | 0xff000008     | r    | r     | Version of core/stick, identitying FPGA bitstream, not the product version or serial. See UDI.                                                                                                                        |
```
Do we talk about 'stick' anywhere else in documentation. Could, should we instead call this Tkey FPGA type and version?

mchack-work replied:
Good point. I just copied the earlier text.

secworks commented on hugo/content/memory/_index.md:
```diff
@@ -35,69 +33,74 @@ The first 2 bits in a 32-bit address.
+| `TK1_CPU_MON_CTRL`  | 0xff000180     | w    | w     | Bit 0 enables CPU execution monitor. Can't be unset. Locks addresses set in `TK1_CPU_MON_FIRST` and `TK1_CPU_MON_LAST`.                                                                                               |
+| `TK1_CPU_MON_FIRST` | 0xff000184     | w    | w     | First adress (XXX word?) of the RAM area monitored for execution attempts.                                                                                                                                            |
+| `TK1_CPU_MON_LAST`  | 0xff000188     | w    | w     | Last address (XXX word?) of the area monitored for execution attempts.                                                                                                                                                |
+
```
Should we add SPI-things her as optional?

mchack-work replied:
The SPI is still an unmerged feature. This is a change to the Dev Handbook to fix bugs in the memory map. 

This will go live on https://dev.tillitis.se/ the moment we merge.

secworks replied:
Ok. Good point.

secworks commented on hugo/content/hw/_index.md:
```diff
@@ -93,10 +93,11 @@ The memory protection is based on two separate mechanisms:
-to the device app and developers don't have to do anything to use it.
+with the contents of the `ADDR_RAM_ASLR` register in the TK1 core
+(addressed with `TK1_RAM_ADDR_RAND` in the memory map). The result is
+used as the RAM address. This is set up by the firmware as part of
```
I suggest that we change both address and data randomization to ADDR_RAM_ADDR_RAND and ADDR_RAM_DATA_RAND. We are performing basically the same randomization operation. for both. Scrambling is possibly more correct now, but is longer. And if move to a PRNG it is randomization.

secworks replied:
Will fix this in a branch for the main Tkey repo.

mchack-work commented:
Wrote some more, fixed some bugs (some things are invisible in app mode and was listed as readable), added some missing stuff, especially in the hardware section. Used the real names from the header file.

Please check again and please check especially for any errors in the description: Is this thing really readable in app mode? Is this thing really an address to a word boundary? Et cetera.

dehanj commented:
Did some tweaking, and added to some things missing in the table. 
Have not yet verified the r, r/w and i.

dehanj reviewed (approved):
Squashed and rebased on main.
